package/kedro_viz/data_access/repositories/modular_pipelines.py

```diff
@@ -165,162 +162,146 @@ def _add_outputs(self, modular_pipeline_id: str, outputs: Set[str]) -> None:
-            # of a top-level modular pipeline as a child to ROOT modular
-            # pipeline
-            self._add_children_to_parent_pipeline(
-                self.tree[ROOT_MODULAR_PIPELINE_ID],
-                modular_pipeline_id,
-                modular_pipeline_inputs_outputs,
+            modular_pipeline.tags.update(node.tags)
+
+            hashed_io_ids = {
+                _hash_input_output(io) for io in set(node.inputs).union(node.outputs)
+            }
+
+            # Compute valid input/output IDs that are not part of the modular pipeline
+            # inputs/outputs or parameters
+            valid_io_ids = hashed_io_ids.difference(
+                modular_pipeline_inputs_outputs, self.parameters
             )
 
+            # Add each valid input/output as a child to the modular pipeline
+            for io_id in valid_io_ids:
+                modular_pipeline.children.add(
+                    ModularPipelineChild(id=io_id, type=GraphNodeType.DATA)
+                )
+                self.node_mod_pipeline_map[io_id].add(modular_pipeline.id)
+
+            self.node_mod_pipeline_map[node_id].add(modular_pipeline.id)
+
     def _add_children_to_parent_pipeline(
         self,
-        parent_node: ModularPipelineNode,
+        modular_pipeline: ModularPipelineNode,
         modular_pipeline_id: str,
         modular_pipeline_inputs_outputs: Set[str],
     ):
         """
-        Helper to add modular_pipeline children correctly to
-        parent modular pipelines in case of nesting.
+        Helper to add modular_pipeline children correctly to parent modular pipelines
+        in case of nesting.
 
         Here we follow the below rules:
-        - A modular pipeline is a child of it's parent modular pipeline
+        - A modular pipeline is a child of its parent modular pipeline
         - Any input/output of a modular pipeline is a child of the parent pipeline
-          unless it is an input/output of that parent pipeline or a parameter dataset
-        - Any input/output of a top-level modular pipeline is added as a child
-          to the root modular pipeline
+        unless it is an input/output of that parent pipeline or a parameter dataset
+        - Any input/output of a top-level modular pipeline is added as a child to
+        the root modular pipeline
 
         Args:
-            parent_node (ModularPipelineNode): The parent modular pipeline node.
+            modular_pipeline (ModularPipelineNode): The modular pipeline node.
             modular_pipeline_id (str): The ID of the modular pipeline to add as a child.
-            modular_pipeline_inputs_outputs (Set[str]): A set of inputs/outputs
-                    to/from the modular pipeline
+            modular_pipeline_inputs_outputs (Set[str]): A set of inputs/outputs to/from
+            the modular pipeline.
         """
+        # Determine the parent modular pipeline ID
+        parent_modular_pipeline_id = (

```

Comment:
I'd simply call this `parent_id`, variables have context and in this function the only thing that can be a parent is a modular pipeline, thus it's implied and doesn't need to be part of the variable name.